examples/chat/README.md

```diff
@@ -5,52 +5,46 @@ This directory holds a very simple chat application.
+room.
 
-- The **room ID** is needed to join that room and participate in chats there;
-  this is printed (as a base-64 string) to allow sharing.
-- The **admin ID** allows destruction of the room instance at the server; the
-  client keeps hold of this ID internally.
+Additionally, an **admin token** allows destruction of the room instance at the
+server; the client keeps hold of this ID internally.
 
-## Command Line Operation
-
-The first client to run typically creates the Oak Application and a first chat
-room inside it:
-
-```bash
-$ ./examples/chat/run
-```
+## Implementation
 
-This will emit a trace line that holds the information needed to:
+The application starts with a "router" node that only ever handles public data,
+and is not allowed to inspect any data protected by a label (e.g. the
+aforementioned authentication bearer token).
 
-- connect to the same Oak application (with `--app_address`)
-- join the chat room (with `--room_id`).
+The router node receives gRPC invocations on an inbound chananel, but only looks
+at the associated label (which is public), and based on the label only, routes
+the request to either an existing or a newly created "room worker node", which
+is created with that specific label and only ever handles data with the same
+label, for the duration of its own lifetime.
 
-```log
-2019-10-24 10:47:20  INFO  chat.cc : 242 : Join this room with --app_address=127.0.0.1:32889 --room_id=NKsceNlg69UbcvryfzmFGnMv9qnZ0DYh6u6gJxujnPPxvHsxMehoD368sumKawVaq9WaSkzrcStoNYLvVNdzhA==
-```
+## Command Line Operation
 
-Another party can then join the same chat room by using these arguments:
+The first client to run typically generates a bearer token for the chat room,
+for instance:
 
 ```bash
-$ ./examples/chat/run -C --app_address=127.0.0.1:32889 --room_id=NKsceNlg69UbcvryfzmFGnMv9qnZ0DYh6u6gJxujnPPxvHsxMehoD368sumKawVaq9WaSkzrcStoNYLvVNdzhA==
+export OAK_CHAT_TOKEN="$(openssl rand -base64 64 | tr -d '\n')"
 ```
 
-Alternatively, another party can create a new chat room running on the same Oak
-Application by just copying the `--app_address` argument, but specifying a new
-room name:
+It then creates the Oak Application and a first chat room inside it, using the
+newly generated bearer token:
 
 ```bash
-$ ./examples/chat/run -C --app_address=127.0.0.1:32889
+./examples/chat/run -C --authorization_token="${OAK_CHAT_TOKEN}"
 ```
 
-This will again emit a trace line with the information needed to join this new
-room (on the same server):
+Another client can then join the same chat room by using the same command.
 
-```log
-2019-10-24 11:04:40  INFO  chat.cc : 242 : Join this room with --app_address=127.0.0.1:32889 --room_id=msSGas1Ie2rtGIvG0bLa2Jh3ODjO35nix46R3j2iYjAcB8zDcJpn/P2DD7c0yB1NMmfoipBSAePJzlXjknm8gg==
-```
+In a more realistic scenario, the first client would securely share the bearer
+token with other participants that it wants to be able to join the chat room,
+using some out-of-band mechanism.
```

Comment:
nit: The existence of a secure out of band mechanism for exchanging bearer tokens means that they probably don't need a chat room to exchange information preivately, so this does not really seem like a more realistic scenario. Perhaps add a comment that this will be improved when public-key options are implemented?

Reply:
I also don't think that we should solve the key distribution problem. The main difference is between exchanging public keys versus private or symmetric keys. For public keys it is just a matter of integrity, not secrecy, so you can use a registry of public keys that you both trust. In the current implementation sharing the bearer token is equivalent to sharing a private or symmetric key. If I already had a secure out of band mechanism to share such sensitive information that provided the needed secrecy I would just send the message itself through this out of band mechanism.